src/librustc_middle/ty/instance.rs

```diff
@@ -502,6 +483,92 @@ impl<'tcx> Instance<'tcx> {
+        Some(substs.as_closure().tupled_upvars_ty())
+    } else if tcx.type_of(def_id).is_generator() {
+        Some(substs.as_generator().tupled_upvars_ty())
+    } else {
+        None
+    };
+    let has_upvars = upvars_ty.map(|ty| ty.tuple_fields().count() > 0).unwrap_or(false);
+    debug!("polymorphize: upvars_ty={:?} has_upvars={:?}", upvars_ty, has_upvars);
+
+    struct PolymorphizationFolder<'tcx> {
+        tcx: TyCtxt<'tcx>,
+    };
+
+    impl ty::TypeFolder<'tcx> for PolymorphizationFolder<'tcx> {
+        fn tcx<'a>(&'a self) -> TyCtxt<'tcx> {
+            self.tcx
+        }
+
+        fn fold_ty(&mut self, ty: Ty<'tcx>) -> Ty<'tcx> {
+            debug!("fold_ty: ty={:?}", ty);
+            match ty.kind {
+                ty::Closure(def_id, substs) => {
+                    let polymorphized_substs = polymorphize(self.tcx, def_id, substs);
+                    self.tcx.mk_closure(def_id, polymorphized_substs)
+                }
+                ty::Generator(def_id, substs, movability) => {
+                    let polymorphized_substs = polymorphize(self.tcx, def_id, substs);
+                    self.tcx.mk_generator(def_id, polymorphized_substs, movability)
+                }
+                _ => ty.super_fold_with(self),
+            }
+        }
+    }
+
+    InternalSubsts::for_item(tcx, def_id, |param, _| {
+        let is_unused = unused.contains(param.index).unwrap_or(false);
+        debug!("polymorphize: param={:?} is_unused={:?}", param, is_unused);
+        match param.kind {
+            // Upvar case: If parameter is a type parameter..
+            ty::GenericParamDefKind::Type { .. } if
+                // ..and has upvars..
+                has_upvars &&
+                // ..and this param has the same type as the tupled upvars..
+                upvars_ty == Some(substs[param.index as usize].expect_ty()) => {
+                    // ..then double-check that polymorphization marked it used..
+                    debug_assert!(!is_unused);
+                    // ..and polymorphize any closures/generators captured as upvars.
+                    let upvars_ty = upvars_ty.unwrap();
+                    let polymorphized_upvars_ty = upvars_ty.fold_with(
+                        &mut PolymorphizationFolder { tcx });
+                    debug!("polymorphize: polymorphized_upvars_ty={:?}", polymorphized_upvars_ty);
+                    ty::GenericArg::from(polymorphized_upvars_ty)
+                },
+
+            // Simple case: If parameter is a const or type parameter..
+            ty::GenericParamDefKind::Const | ty::GenericParamDefKind::Type { .. } if
+                // ..and is within range and unused..
+                unused.contains(param.index).unwrap_or(false) =>
+                    // ..then use the identity for this parameter.
+                    tcx.mk_param_from_def(param),
+
+            // Otherwise, use the parameter as before.
+            _ => substs[param.index as usize],
```

Comment:
can't we just always polymorphize here ?
```suggestion
            _ => {
                let polymorphized_ty = substs[param.index as usize].fold_with(
                        &mut PolymorphizationFolder { tcx });
                ty::GenericArg::from(polymorphized_ty)
            },
```

Reply:
this would fix this while also optimizing things like

```rust
fn test<T>() {
    [1, 2, 3].iter().map(|_| ()).collect::<Vec<()>>();
}

fn main() {
    test::<String>();
    test::<Vec<String>>();
}
```

Reply:
It would require us to also change that early exit here afaict.
```rust
    if unused.is_empty() {
        // Exit early if every parameter was used.
        return substs;
    }
```

Reply:
Applying this suggestion would also work (instead of adding another arm like the PR currently does). I avoided doing this because I figured it would be more expensive (traversing every substitution rather than just upvars). 

That said, I might be missing something, because I’m now sure how it would help with the example in your second comment (`test` and its closure would be polymorphised now and with your change, no?).

Reply:
My concern is the following: https://rust.godbolt.org/z/fGeseM

We currently create 2 instances of `slice::Iter::map` here, while after this change we probably only create one

Reply:
And I think this should actually happen somewhat frequently with iterator adapters in generic functions

Reply:
Fixed!